File: cloudflare/cloudflare-dns.php
#!/usr/bin/php
<?php
/**
 * Script by Phantium
 */

class CloudFlareDNS {
        private $api_url, $api_params, $request_url;

        private $username = "cloudflare_username";
        private $apikey = "cloudflare_apikey";

        function getIP() {
                $ip_url = file_get_contents('http://ip.raw.re');
                $this->ip = $ip_url;
                if (strlen($this->ip) < 10) {
                        return false;
                } else {
                        return $this->ip;
                }
        }

        function setIP($hostname) {
                if ($this->getIP() === false) {
                        echo "Error: IP was not succesfully retrieved.";
                        return;
                }
                $this->api_url = "https://www.cloudflare.com/api_json.html";
                $this->api_params = "?a=DIUP&hosts=". $hostname ."&u=". $this->username .
                                    "&tkn=". $this->apikey ."&ip=". $this->getIP();
                $this->request_url = $this->api_url . $this->api_params;
                try {
                        $this->result = file_get_contents($this->request_url);
                }
                catch (Exception $e) {
                        echo "ERROR, Caught exception:", $e->getMessage(), "\n";
                        return;
                }
                return $this->result."\n";
        }
}

$cf = new CloudFlareDNS;
echo $cf->setIP('dns.yourdomain.com');

File: cloudflare/cloudflare-dns.py
#!/usr/bin/env python
# Script by Phantium

import sys, urllib2, socket

ip_url_address = "http://ip.raw.re" # Website providing IP lookup, can be anything as long it outputs just the IP.
username = "your@email.com" # Username for cloudflare
apikey = "yourapikey" # API key for cloudflare

def getIP():
    try:
        ip = urllib2.urlopen(ip_url_address).read()
    except URLError as e:
        print e.reason
    except:
        print "Unable to retrieve IP."
    else:
        return ip

def checkIP(host):
    current_ip = socket.getaddrinfo(host, None, 0)[0][4][0]
    new_ip = getIP()
    if not current_ip == new_ip:
        print "IP changed, updating..."
        return new_ip
    else:
        print "IP: '%s' is already defined for '%s', nothing to do." % (current_ip,host)

def setIP(hostname):
    check_ip = checkIP(hostname)
    if check_ip:
        api_url = "https://www.cloudflare.com/api_json.html"
        api_params = ("?a=DIUP&hosts=%s&u=%s&tkn=%s&ip=%s" % (hostname, username, apikey, check_ip))
        request_url = api_url + api_params
        print "IP for '%s' set to: '%s'" % (hostname, check_ip)
        try:
            urllib2.urlopen(request_url).read()
        except URLError as e:
            print e.reason
        except:
            print "Something broke but I don't know why!"

setIP("dns.yourdomain.com")

